src/passes/inlineAssemblyTransformer.ts

```diff
@@ -0,0 +1,239 @@
+import { ASTMapper } from '../ast/mapper';
+import { AST } from '../ast/ast';
+import {
+  createExpressionStatement,
+  createIdentifier,
+  createNumberLiteral,
+  createTuple,
+} from '../utils/nodeTemplates';
+
+import {
+  InlineAssembly,
+  YulNode,
+  Identifier,
+  ASTNode,
+  BinaryOperation,
+  Assignment,
+  Block,
+  Expression,
+  ExpressionStatement,
+  VariableDeclaration,
+  UncheckedBlock,
+  Literal,
+  Statement,
+  ASTContext,
+  FunctionCall,
+  FunctionCallKind,
+} from 'solc-typed-ast';
+import assert from 'assert';
+import { getErrorMessage, WillNotSupportError } from '../utils/errors';
+
+/* 
+TODO: Research the difference between div, sdiv, mod, smod, shr, sar
+One scenario when the result is different:
+int8 r;
+int8 a = -1;
+int8 b = 2;
+assembly {
+  r := div(a, b)
+}
+
+div returns -1 while sdiv returns 0 
+*/
+const binaryOps: { [name: string]: string } = {
+  add: '+',
+  mul: '*',
+  div: '/',
+  sub: '-',
+  shr: '>>',
+  shl: '<<',
+  mod: '%',
+  exp: '**',
+};
+// Consider grouping functions by number of arguments. They probbably all use uint256 with the only difference being number of args.
+const solidityEquivalents: { [name: string]: string } = {
+  addmod: 'function (uint256,uint256,uint256) pure returns (uint256)',
+  mulmod: 'function (uint256,uint256,uint256) pure returns (uint256)',
+};
+
+function generateTransformerName(nodeType: string) {
+  return `transform${nodeType}`;
+}
+
+class YulVerifier {
+  errors: string[] = [];
+
+  verifyNode(node: YulNode) {
+    const transformerName = generateTransformerName(node.nodeType);
+    if (!(transformerName in YulTransformer.prototype))
+      this.errors.push(`${node.nodeType} is not supported`);
+
+    const verifierName = `check${node.nodeType}`;
+    const method = this[verifierName as keyof YulVerifier] as (node: YulNode) => void;
+    if (method !== undefined) return method.bind(this)(node);
+  }
+  checkYulAssignment(node: YulNode) {
+    this.verifyNode(node.value);
+  }
+
+  checkYulFunctionCall(node: YulNode) {
+    if (
+      binaryOps[node.functionName.name] === undefined &&
+      solidityEquivalents[node.functionName.name] === undefined
+    )
+      this.errors.push(`${node.functionName.name} is not supported`);
+    node.arguments.forEach((arg: YulNode) => this.verifyNode(arg));
+  }
+}
+
+class YulTransformer {
+  vars: Map<string, VariableDeclaration>;
+  ast: AST;
+
+  constructor(ast: AST, externalReferences: any[], context: ASTContext) {
```

Comment:
Is it possible to specify what `externalReference` type is going to be. Or what do you expect it to be?

Reply:
A more precise type is only possible after filtering which is done in the constructor currently. The type is the same as in [InlineAssembly.externalReferences](https://github.com/ConsenSys/solc-typed-ast/blob/master/src/ast/implementation/statement/inline_assembly.ts#L12)